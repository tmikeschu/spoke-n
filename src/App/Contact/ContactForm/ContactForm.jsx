import React, { Component } from "react"
import APIService from "../../APIService/APIService"
import "../../App.css"

const apiService = new APIService("https://spoken-api.herokuapp.com")

export default class ContactForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
      contact: {
        name: "",
        email: "",
        message: ""
      },
      success: false,
      failure: false,
      error: ""
    }
    this.emptyForm = this.state.contact

    this.handleChange = this.handleChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleChange(event) {
    this.setState({
      ...this.state,
      contact: {
        ...this.state.contact,
        [event.target.name]: event.target.value
      }
    })
  }

  messageData = () => ({
    contact: this.state.contact
  })

  async handleSubmit(event, service, messageData) {
    event.preventDefault()
    const response = await service.post("api/v1/contact", messageData)
    this.handleResponse(response)
    return response
  }

  handleResponse(response) {
    this.setState(this.stateFor(response))
    setTimeout(() => {
      this.setState({
        success: false,
        failure: false
      })
    }, 3000)
  }

  stateFor = (response) => {
    const { status, data } = response
    return (
      status === 201 && { success: true, contact: this.emptyForm }
    ) || (
      status === 400 && { failure: true, error: String(data) }
    )
  }

  render() {
    return (
      <article className="component-container contact-form">
        <form onSubmit={event => this.handleSubmit(event, apiService, this.messageData())}>
          <input 
            required type="text" name="name" value={this.state.contact.name} 
            onChange={this.handleChange} placeholder="Your Name"
          />

          <input 
            required type="text" name="email" value={this.state.contact.email} 
            onChange={this.handleChange} placeholder="Email Address"
          />

          <input type="hidden" name="_subject" value="Website contact" />

          <textarea 
            required name="message" value={this.state.contact.message} 
            onChange={this.handleChange} placeholder="What's up?"
          />

          <input type="submit" value="Send" />

          <p style={{ display: this.state.success ? "" : "none" }}>
            Message Sent!
          </p>
          <p style={{ display: this.state.failure ? "" : "none" }}>
            Message failure: { this.state.error }
          </p>
        </form>
        <p>
          *Please allow ample time for a response as we are busy bicycling 15,000 miles.
        </p>
      </article>
    )
  }
}

