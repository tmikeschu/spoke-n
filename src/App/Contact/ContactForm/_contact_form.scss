.contact-form {
  width: 100%;
  form {
    display: flex;
    flex-flow: column nowrap;
    width: 50%;
    margin: 0 auto;
    textarea, input {
      width: 100%;
      margin: 8px auto;
      font-size: 16px;
      padding: 16px;
      box-sizing: border-box;
    }
    input {
      height: 32px;
    }
    textarea {
      resize: none;
      height: 96px;
    }
    input {
      height: auto;
    }
  }
  p {
    text-align: center;
    margin: 8px auto;
  }

  @media (max-width: 640px) {
    width: 100%;
    form {
      width: 100%;
    }
    p {
      text-align: left;
    }
  }
}
