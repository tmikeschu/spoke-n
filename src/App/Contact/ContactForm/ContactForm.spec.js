import React from "react"
import { shallow, mount } from "enzyme"
import ContactForm from "./ContactForm"

const fakeService = {
  post(url, data) {
    return {
      status: 201,
      data: "AHH!"
    }
  }
}
const fakeResponse1 = {
  status: 201,
  data: "COOL"
}
const fakeResponse2 = {
  status: 400,
  data: "AHH!"
}
const expectedMessageData = {
  contact: {
    name: "Tommy", email: "tom@my.crosby", message: "yo yo yo"
  }
}
const event = {
  contact: {
    name: "Tommy", email: "tom@my.crosby", message: "yo yo yo"
  },
  preventDefault() {}
}

describe("<ContactForm />", () => {
  it("renders a contact form", () => {
    const contactForm = shallow(<ContactForm />)
    expect(contactForm.find("article.contact-form").length).toEqual(1)
    expect(contactForm.find("form").length).toEqual(1)
    expect(contactForm.find("form input[name='name']").length).toEqual(1)
    expect(contactForm.find("form input[name='email']").length).toEqual(1)
    expect(contactForm.find("form textarea").length).toEqual(1)
    expect(contactForm.find("form input[name='_subject']").length).toEqual(1)
    expect(contactForm.find("form input[type='submit']").length).toEqual(1)
    expect(contactForm.find("article p").length).toEqual(3)
  })

  describe("#handleChange", () => {
    it("is called on a key change", () => {
      const restore = ContactForm.prototype.handleChange
      const mock = ContactForm.prototype.handleChange = jest.fn()
      const contactForm = mount(<ContactForm />)
      contactForm.find(".contact-form input[name='name']").simulate("change")
      expect(mock).toHaveBeenCalledTimes(1)
      contactForm.find(".contact-form input[name='email']").simulate("change")
      expect(mock).toHaveBeenCalledTimes(2)
      
      ContactForm.prototype.handleChange = restore
    })

    it("updaates state values", () => {
      const contactForm = mount(<ContactForm />)
      const nameField = contactForm.find(".contact-form input[name='name']")

      expect(nameField.node.value).toEqual('')

      nameField.node.value = "a"
      nameField.simulate("change", nameField)

      expect(nameField.node.value).toEqual('a')
    })
  })

  describe("#messageData", () => {
    it("returns an object of existing state", () => {
      const contactForm = shallow(<ContactForm />)
      contactForm.setState(expectedMessageData)
      const messageData = contactForm.instance().messageData()

      expect(messageData).toMatchObject(expectedMessageData)
    })
  })

  describe("#handleSubmit", () => {
    it("is called on submit", () => {
      const restore = ContactForm.prototype.handleSubmit
      const mock = ContactForm.prototype.handleSubmit = jest.fn()
      const contactForm = mount(<ContactForm />)
      contactForm.find(".contact-form form").simulate("submit")

      expect(mock).toHaveBeenCalled()
      ContactForm.prototype.handleSubmit = restore
    })

    it("makes a post request", async () => {
      const contactForm = shallow(<ContactForm />)
      const response = await contactForm.instance()
        .handleSubmit(event, fakeService, expectedMessageData)

      expect(response).toEqual(response)
    })
  })

  describe("#handleResponse", () => {
    it("is called on submit", async () => {
      const restore = ContactForm.prototype.handleResponse
      const mock = ContactForm.prototype.handleResponse = jest.fn()
      const contactForm = shallow(<ContactForm />)
      await contactForm.instance()
        .handleSubmit(event, fakeService, expectedMessageData)

      expect(mock).toHaveBeenCalled()
      ContactForm.prototype.handleResponse = restore
    })

    it("sets state for a 201", () => {
      const contactForm = shallow(<ContactForm />).instance()
      expect(contactForm.state.success).toBeFalsy()
      contactForm.handleResponse(fakeResponse1)
      expect(contactForm.state.success).toBeTruthy()
    })

    it("sets state for a 400", () => {
      const contactForm = shallow(<ContactForm />).instance()
      expect(contactForm.state.failure).toBeFalsy()
      contactForm.handleResponse(fakeResponse2)
      expect(contactForm.state.failure).toBeTruthy()
    })

    it("and state times out", () => {
      const contactForm = shallow(<ContactForm />).instance()
      contactForm.handleResponse(fakeResponse1)
      expect(contactForm.state.success).toBeTruthy()
      setTimeout(() => {
        expect(contactForm.state.success).toBeFalsy()
      }, 3000)
    })

    it("and state times out", () => {
      const contactForm = shallow(<ContactForm />).instance()
      contactForm.handleResponse(fakeResponse2)
      expect(contactForm.state.failure).toBeTruthy()
      setTimeout(() => {
        expect(contactForm.state.failure).toBeFalsy()
      }, 3000)
    })
  })

  describe("#stateFor", () => {
    it("returns success and contact for 201", () => {
      const contactForm = shallow(<ContactForm />).instance()
      const response = contactForm.stateFor(fakeResponse1)
      expect(response).toMatchObject({
        success: true, contact: { name: "", email: "", message: ""}
      })
    })

    it("returns success and contact for 400", () => {
      const contactForm = shallow(<ContactForm />).instance()
      const response = contactForm.stateFor(fakeResponse2)
      expect(response).toMatchObject({
        failure: true, error: "AHH!"
      })
    })
  })
})

