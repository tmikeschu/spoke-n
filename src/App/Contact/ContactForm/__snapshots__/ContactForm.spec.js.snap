// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactForm /> snapshot is valid 1`] = `
Array [
  <article
    className="component-container contact-form"
>
    <form
        onSubmit={[Function]}
    >
        <input
            name="name"
            onChange={[Function]}
            placeholder="Your Name"
            required={true}
            type="text"
            value=""
        />
        <input
            name="email"
            onChange={[Function]}
            placeholder="Email Address"
            required={true}
            type="text"
            value=""
        />
        <input
            name="_subject"
            type="hidden"
            value="Website contact"
        />
        <textarea
            name="message"
            onChange={[Function]}
            placeholder="What's up?"
            required={true}
            value=""
        />
        <input
            type="submit"
            value="Send"
        />
        <p
            style={
                Object {
                    "display": "none",
                  }
            }
        >
            Message Sent!
        </p>
        <p
            style={
                Object {
                    "display": "none",
                  }
            }
        >
            Message failure: 
            
        </p>
    </form>
    <p>
        *Please allow ample time for a response as we are busy bicycling 15,000 miles.
    </p>
</article>,
]
`;
